Enable TLS Support for C/C++

gyrovorbis opened the pull request:
This is the work required to support static TLS for KOS and in the toolchains per the SH TLS ABI is variant I.

- [X] Remove disabled TLS in the toolchain build scripts
- [X] Add "tcbhead" intrusive structure to KOS's thread structure, which is what's stored within GBR
- [X] Allocating extra chunk of memory for "tcbhead" followed by .tdata and .tbss segements
- [X] Set static TLS block data based on .tdata values for initialized data, 0 for .tbss data
- [X] Added TLS-specific section exports to the linker script for access within KOS
- [X] Created example flexing TLS using thread-local variables requiring both .tbss and .tdata initialization
- [X] Tested with GCC12.2.0
- [X] Tested with GCC9.3.0
- [X] Tested with GCC4.7.0
- [X] Tested with GCC13.1.0 
- [X] TLS variables working properly for the main thread
- [X] Manually aligned TLS variables working properly 

This implementation still provides a void *dtv field which is NULL initialized for now, which can later be used to add support for non-static dynamic module TLS in the future as well. 

This PR also adds an example of using TLS storage that also doubles as a test suite to flex the hell out of this implementation and return EXIT_SUCCES/EXIT_FAILURE based on the results (which we can then use in CI/automation once we get return values in DC-Tool).

cepawiel commented:
This should be in a working state if anyone wants to help test. I've tested it working on all three (GCC4.7, GCC9, GCC12) toolchains. While I can't test on NAOMI (yet) I believe the changes to the linker script are pretty sane and copy/pasted from the tested Dreamcast one and should work.


> Ensured Newlib's C and C++ stdlibs are now using TLS gracefully for reentrancy

I think any Newlib changes should be investigated in a future PR instead of this one, as Newlib support for TLS replacing the reent struct is relatively new. It was added in [this](https://sourceware.org/git/?p=newlib-cygwin.git;a=commit;h=ea99f21ce6c8a332845439ea87f792dbaf679d42) commit, and would only be possible with the GCC12 toolchain.

gyrovorbis commented:
> > Ensured Newlib's C and C++ stdlibs are now using TLS gracefully for reentrancy
> 
> I think any Newlib changes should be investigated in a future PR instead of this one, as Newlib support for TLS replacing the reent struct is relatively new. It was added in [this](https://sourceware.org/git/?p=newlib-cygwin.git;a=commit;h=ea99f21ce6c8a332845439ea87f792dbaf679d42) commit, and would only be possible with the GCC12 toolchain.
Yeah, after talking to you, I agree. Updated the main comment to remove that but also added the diligent toolchain testing you've done too. When I get a chance, I'll help do some testing on the C++ side.

gyrovorbis commented:
If anyone has anything they'd like to add or see, this PR is basically ready other than a bit more testing, which I'm holding off on until I finish writing a bunch of unit tests for libGimbal's concurrency stuff, which is about to flex this all pretty hard!

gyrovorbis commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,75 @@
+    for(i = 0; i < thread_count; i++) {
+        threads[i] = thd_create(0, thd, (void *) i);
+    };
+
```
Adding the following code here for accessing the TLS data also from the main thread will immediately illustrate one of the two remaining issues. The main thread's TLS data and pointers seem to be incorrect:
```
// ========== DO IT IN THE MAIN THREAD TOO ==========
    for (i = 0; i < 5; i++){        
        printf("Thread[MAIN]\tbss_test = 0x%lX\n", tbss_test);
        tbss_test++;
        thd_sleep(50);
    }

    for (i = 0; i < 5; i++){
        printf("Thread[MAIN]\ttdata_test = 0x%lX\n", tdata_test);
        tdata_test++;
        thd_sleep(50);
    }
// =================================================
```

gyrovorbis commented:
I have fantastic news to report on the quest for TLS... I CANNOT REPRODUCE THE PROBLEMS I FOUND ON GCC13! I'm not sure what the heck is going on or if 12 had issues or what the deal was, but I've modified your example to flex the two things I was mentioning 1) main thread using TLS data 2) weirdly aligned TLS data, and it all works fine.

I am, however, still facing one last issue with my libGimbal codebase using TLS... Every time I try to link to a statically linked library that uses TLS variables that get referenced from the main program, the compiler emits `__emu_tls_()` symbols, which seem to totally break the code in the final binary... I'm trying to play with toolchain configurations and global build flags (stuff like `-fno-pic` and `-fno-pie` to see if I can get it to always emit GBR offset accesses to TLS variables.

Btw, is there any way you can give me permission to push back to this branch? I can add my changes to your example flexing the other two issues so you can verify they're working for you too.

cepawiel commented:
I was able to replicate the issue with using TLS on the main thread. I just wasn't setting the GBR register for the main thread so it was not accessing the correct memory region when attempting to access TLS variables. This was only an issue for the main thread as it's the only thread that sets up its own thread structure, all others will have the GBR loaded during a context switch.

gyrovorbis commented:
It has been a long long long time coming, but we got TLS working!!!

It has been validated with both the new `compiler_tls` example from Colto and my MASSIVE `libGimbal` unit test suite. Here's the threading tests I put KOS through that all passed:
https://github.com/gyrovorbis/libgimbal/blob/master/test/source/core/gimbal_thread_test_suite.c

Between the two, I feel like we've thrown everything at the TLS implementation that I can think of... she's finally ready! Really excited about this one!

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/Makefile:
```diff
@@ -0,0 +1,31 @@
+#
+# basic/threading/compiler_tls/Makefile
+#
+# (c)2001 Megan Potter
+# (c)2023 Colton Pawielski
```
```suggestion
# Copyright (C) 2001 Megan Potter
# Copyright (C) 2023 Colton Pawielski
```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+
+   compiler_tls.c
+
+   (c) 2023 Colton Pawielski
+   (c) 2023 Falco Girgis
```
```suggestion
   Copyright (C) 2023 Colton Pawielski
   Copyright (C) 2023 Falco Girgis
```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+
+    /* Ensure zero-initialized .TBSS data have the correct initial 
+       values and are unique to each thread. */
+    for (i = 0; i < 5; i++){        
```
```suggestion
    for(i = 0; i < 5; i++) {
```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+
+    /* Ensure value-initialized .TDATA data have the correct initial 
+       values and are unique to each thread. */
+    for (i = 0; i < 5; i++){
```
```suggestion
    for(i = 0; i < 5; i++) {
```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    /* Ensure default-aligned .TBSS data is initialized properly. */
+    for(int i = 0; i < 256; ++i)
+        if(tls_uint16[i] != 0) {
+            fprintf(stderr, "tls_uint16[%d] failed!\n", i);
+            ret = -1;
+            break;
+        }
```
```suggestion
    for(i = 0; i < 256; ++i) {
        if(tls_uint16[i] != 0) {
            fprintf(stderr, "tls_uint16[%d] failed!\n", i);
            ret = -1;
            break;
        }
    }
```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+    bool reproduced = false;
+
+    printf("[");
+    for (int i = 0; i < 3; i++) {
```
```suggestion
    for(i = 0; i < 3; i++) {
```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+
+    printf("[");
+    for (int i = 0; i < 3; i++) {
+        if (tls_buff4.inner[i] != 2) 
```
```suggestion
        if(tls_buff4.inner[i] != 2) 
```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+            reproduced = true;
+        
+        printf("%d, ", tls_buff4.inner[i]);
+    }
```
```suggestion
    }

```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+    printf("]\n");
+
+    printf("[");
+    for (int i = 0; i < 3; i++) {
```
```suggestion
    for(i = 0; i < 3; i++) {
```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+
+    printf("[");
+    for (int i = 0; i < 3; i++) {
+        if (tls_buff16.inner[i] != 1) 
```
```suggestion
        if(tls_buff16.inner[i] != 1) 
```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+            reproduced = true;
+        
+        printf("%d, ", tls_buff16.inner[i]);
+    }
```
```suggestion
    }

```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+    }
+    printf("]\n");
+
+    if (reproduced) {
```
```suggestion
    if(reproduced) {
```

ljsebald commented on examples/dreamcast/basic/threading/compiler_tls/compiler_tls.c:
```diff
@@ -0,0 +1,199 @@
+    /* Create a bunch of threads and put each through 
+       the same series of tests on their own (hopefully)
+       independent set of thread-local variables. */
+    for (i = 0; i < thread_count; i++)
+        threads[i] = thd_create(0, thd, (void *)i + 1);
```
```suggestion
    for(i = 0; i < thread_count; i++) {
        threads[i] = thd_create(0, thd, (void *)i + 1);
    }
```

ljsebald commented on kernel/thread/thread.c:
```diff
@@ -327,6 +342,82 @@ int thd_remove_from_runnable(kthread_t *thd) {
+   followed by .TBSS, very carefully ensuring alignment of each
+   subchunk. 
+*/
+static void* thd_create_tls_data(void) {
```
```suggestion
static void *thd_create_tls_data(void) {
```

ljsebald replied:
Please move all variables to the top of the function. In-kernel we stick with C89/C90-style.

Also, please replace aligned_alloc with memalign since we don't compile the main kernel as C11. `<malloc.h>` is already included, so the function should be prototyped already.

gyrovorbis commented:
![Screenshot from 2023-09-15 20-53-04](https://github.com/KallistiOS/KallistiOS/assets/5545520/237ec436-d947-454a-88b0-491f100cf7a5)
Review feedback addressed, rebuilt KOS along with the example and nothing has broken (thank god). :)
